File: summer2019/week3/lemonade-change.java
// Uses the multiset class in Multiset.java
class Solution {
    public boolean lemonadeChange(int[] bills) {
        Multiset<Integer> change = new Multiset<>();
        for (int bill : bills) {
            if (!executeTransaction(change, bill)) {
                return false;
            }
        }
        return true;
    }
    
    public boolean executeTransaction(Multiset<Integer> changeOnHand, int bill) {
        // Take customer's bill and add it to change on hand
        changeOnHand.add(bill);
        
        // Check if customer paid exact change
        int changeNeeded = bill - 5;
        if (changeNeeded == 0) {
            return true;
        }
        
        int[] options = {20, 10, 5};
        int p = 0;
        while (changeNeeded != 0) {
            // Return false if no options available
            if (p >= options.length) {
                return false;
            }
            
            // Set p to the highest possible bill
            if (options[p] > changeNeeded || !changeOnHand.contains(options[p])) {
                p++;
                continue;
            }
            
            changeNeeded -= options[p];
            changeOnHand.remove(options[p]);
        }
        return true;
    }
}

File: 2026/week3/lemonade-change.py
class Solution:
    def lemonadeChange(self, bills: List[int]) -> bool:
        amounts = [20, 10, 5]
        change = {
            20: 0,
            10: 0,
            5: 0
        }
        for bill in bills:
            change[bill] += 1
            change_needed = bill - 5
            i = 0
            while change_needed > 0:
                if i >= len(amounts):
                    return False
                elif amounts[i] <= change_needed and change[amounts[i]] > 0:
                    change[amounts[i]] -= 1
                    change_needed -= amounts[i]
                else: 
                    i += 1
        return True
